File: linkedlist/Q143.java
public class Q143 {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) return;

        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode mid = slow;
        ListNode reversed = reverseList(mid.next);
        mid.next = null;

        ListNode ptr = head, next;
        while (reversed != null) {
            next = ptr.next;
            ptr.next = reversed;
            reversed = reversed.next;
            ptr.next.next = next;
            ptr = next;
        }
    }

    private ListNode reverseList(ListNode head) {
        ListNode reversed = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = reversed;
            reversed = head;
            head = temp;
        }
        return reversed;
    }
}

File: LinkedList-Stack-Queue/Q143.py
# https://leetcode.com/problems/reorder-list/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return head

        slow = fast = ListNode(0, head)
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
        mid = slow
        right = mid.next
        mid.next = None

        reversed = None
        while right:
            next = right.next
            right.next = reversed
            reversed = right
            right = next
        right = reversed

        left = head
        while right:
            nextLeft = left.next

            left.next = right
            right = right.next
            left.next.next = nextLeft
            left = nextLeft
        return head
